src/toy-qvm.lisp

```diff
@@ -0,0 +1,158 @@
+(in-package #:qvm)
+
+(defclass toy-qvm (pure-state-qvm)
+  ((readout-povms
+    :initarg :readout-povms
+    :accessor readout-povms
+    :initform (make-hash-table :test 'eql)
+    :documentation "Noisy readout encoded as diagonal single qubit
+    POVM given as a 4-element list (p(0|0) p(0|1) p(1|0) p(1|1))")))
+
+
+;; TODO declare inlinabldee, turn on optimization
+(defun perturb-measurement (actual-outcome p00 p01 p10 p11)
+  "Given the readout error encoded in the POVM (see documentation of NOISY-QVM)
+randomly sample the observed (potentially corrupted) measurement outcome."
+ ; (format t "actual: ~a p0|0: ~a p1|0: ~a" actual-outcome p00 p10)
+  (check-type actual-outcome bit)
+  (check-probabilities p00 p01 p10 p11)
+  (let ((r (random 1.0d0)))
+    (ecase actual-outcome
+      ((0) (if (<= r p00) 0 1))
+      ((1) (if (<= r p01) 0 1)))))
+
+
+(defgeneric set-readout-povm (qvm qubit povm)
+  (:documentation "For a QUBIT belonging to a QVM specify a POVM to encode
+possible readout errors. POVM must be a 4-element list of double-floats."))
+
+
+(defun check-probabilities (p00 p01 p10 p11)
+  "Make sure that the probabilities passed into the POVM are all between 0.0 and 1.0"
+  (check-type p00 (double-float 0.0d0 1.0d0))
+  (check-type p01 (double-float 0.0d0 1.0d0))
+  (check-type p10 (double-float 0.0d0 1.0d0))
+  (check-type p11 (double-float 0.0d0 1.0d0)))
+
+
+(defun print-assignment-probs (qvm q)
+  (let* ((povm-map (readout-povms qvm))
+         (povm (gethash q povm-map)))
+    (destructuring-bind (p00 p01 p10 p11) povm
+      (format t "probabilities: p0|0 : ~a,  p0|1 : ~a,  p1|0 : ~a, p1|1 : ~a" p00 p01 p10 p11))))
+
+
+(defun measure-bit-flip-noise (qvm q numshots)
+  (let ((ones 0)
+        (program "DECLARE R0 BIT; X 0; H 0; MEASURE 0 R0"))
+    (loop :repeat numshots
+       :do (incf ones (do-noisy-measurement qvm q program)))
+    (format t "result: ~a" (float (/ ones numshots)))))
+
+
+(defun estimate-assignment-probabilities (qvm q numshots)
+  (let ((results-i 0)
+        (results-x 0)
+        (program1 "DECLARE R0 BIT; I 0; MEASURE 0 R0")
+        (program2 "DECLARE R0 BIT; X 0; MEASURE 0 R0"))
+    (loop :repeat numshots
+      :do (incf results-i (do-noisy-measurement qvm q program1))
+      :do (incf results-x (do-noisy-measurement qvm q program2))
+          )
+    (let ((p00 (- 1 (float (/ results-i numshots))))
```

Comment:
If you define `results-i` as a float (either with `0.0` or `0d0` depending on your need) then you can get rid of the explicit casting.